File: php/tests/TraceTest.php
<?php

declare(strict_types=1);

namespace Xag\FlightRecorder\Tests;

use PHPUnit\Framework\TestCase;
use Xag\FlightRecorder\Trace;
use Xag\FlightRecorder\Tracer;
use Xag\FlightRecorder\TraceRun;

/** Variable-level tracing: every local, on every executed line. */
final class TraceTest extends TestCase
{
    private const SUBJECT = 'Xag\\FlightRecorder\\Traced\\TracedToy';

    private function source(): string
    {
        return __DIR__ . '/resources/TracedToy.php';
    }

    private function grade(array $answers): TraceRun
    {
        return Tracer::run([$this->source()], self::SUBJECT, 'gradePercent', $answers);
    }

    /** Tracing must not change the answer, including when the answer is wrong. */
    public function testTheInstrumentedCopyStillParsesAndStillComputes(): void
    {
        $run = $this->grade([1, 1, -1, 0]);
        self::assertSame(66, $run->result);
        self::assertFalse($run->trace->isEmpty());
    }

    /**
     * Values are DATA, not renderings — this is the whole point of trace version 2, and it is
     * what lets a claim do arithmetic instead of string matching.
     */
    public function testAVariableTimelineIsALookupNotAnInference(): void
    {
        $t = $this->grade([1, 1, -1, 0])->trace;

        self::assertSame(3, $t->last('answered')->value);
        self::assertSame(2, $t->last('correct')->value);
        self::assertSame(66, $t->last('pct')->value);
        self::assertIsInt($t->last('answered')->value);
    }

    public function testTheArgumentsAVariableArrivedWithAreRecorded(): void
    {
        $t = $this->grade([1, 1, -1, 0])->trace;
        $calls = $t->calls('gradePercent');

        self::assertCount(1, $calls);
        self::assertSame([1, 1, -1, 0], $calls[0]['args']['answers']);
    }

    public function testAReturnIsObserved(): void
    {
        $t = $this->grade([1, 1, -1, 0])->trace;
        $returns = $t->returns('gradePercent');

        self::assertCount(1, $returns);
        self::assertSame(66, $returns[0]['value']);
    }

    /**
     * The reason the feature exists.
     *
     * The result passes every plausible check about itself — an integer, between 0 and 100 — and
     * is still wrong. No claim about the RESULT can catch this. The claim that catches it is
     * about an internal variable.
     */
    public function testTheBugIsCondemnedByItsOwnTrace(): void
    {
        $answers = [1, 1, -1, 0];
        $run = $this->grade($answers);

        self::assertGreaterThanOrEqual(0, $run->result);
        self::assertLessThanOrEqual(100, $run->result);

        $answered = $run->trace->last('answered')->value;
        self::assertNotSame(
            count($answers),
            $answered,
            'the percentage is computed over questions ANSWERED, not questions ASKED'
        );
        self::assertSame(3, $answered);
    }

    public function testAnExceptionIsObservedWithTheStateUpToTheThrow(): void
    {
        $threw = false;
        try {
            Tracer::run([$this->source()], self::SUBJECT, 'divide', 7, 0);
        } catch (\DivisionByZeroError) {
            $threw = true;
        }
        self::assertTrue($threw, 'the exception must propagate untouched');

        $run = Tracer::run([$this->source()], self::SUBJECT, 'divide', 7, 2);
        self::assertSame(35, $run->result);
        self::assertSame(70, $run->trace->last('scaled')->value);
    }

    public function testAnUntracedProcessAnswersNeverObservedRatherThanPassingVacuously(): void
    {
        $t = Trace::empty();

        self::assertTrue($t->isEmpty());
        self::assertSame([], $t->values('answered'));
        self::assertNull($t->last('answered'));
        self::assertSame('answered: never observed', $t->render('answered'));
    }

    public function testAVersionOneTraceIsRefusedRatherThanHalfUnderstood(): void
    {
        $this->expectExceptionMessageMatches('/older tracer/');
        Trace::parse('{"e":"H","trace_version":1}' . "\n" . '{"e":"L","fn":"f","at":"x:1","d":{}}');
    }

    public function testTheTraceRoundTripsThroughItsArtifactForm(): void
    {
        $t = $this->grade([1, 1, -1, 0])->trace;
        $again = Trace::parse($t->toJsonl());

        self::assertSame($t->size(), $again->size());
        self::assertSame($t->last('pct')->value, $again->last('pct')->value);
    }

    /**
     * A location past the end of the real file would mean the trace is pointing at a file that
     * exists nowhere on the reader's disk.
     */
    public function testLocationsPointAtTheOriginalFileNotTheInstrumentedOne(): void
    {
        $t = $this->grade([1, 1, -1, 0])->trace;
        $realLines = count(file($this->source()));

        self::assertNotEmpty($t->events);
        foreach ($t->events as $e) {
            $at = (string) $e['at'];
            self::assertStringStartsWith('TracedToy.php:', $at);
            $line = (int) substr($at, strlen('TracedToy.php:'));
            self::assertGreaterThan(0, $line);
            self::assertLessThanOrEqual($realLines, $line, "location $at is past the end of the file");
        }
    }

    /** The delta belongs to the statement that produced it, not the one about to run. */
    public function testADeltaIsReportedAtTheStatementThatCausedIt(): void
    {
        $t = $this->grade([1, 1, -1, 0])->trace;
        $src = file($this->source());

        $obs = $t->first('pct');
        self::assertNotNull($obs);
        $line = (int) substr($obs->at, strlen('TracedToy.php:'));
        self::assertStringContainsString(
            '$pct = intdiv',
            $src[$line - 1],
            'the observation of $pct must name the line that assigned it'
        );
    }
}

File: java/src/test/java/io/github/xag/flightrecorder/TraceTest.java
package io.github.xag.flightrecorder;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Variable-level tracing, end to end: rewrite the source, compile it to memory, run it, and read
 * what it believed while it ran.
 */
class TraceTest {

    private static final String TOY = "src/test/resources/traced/TracedToy.java";

    @Test
    void theInstrumentedCopyStillCompilesAndStillComputes() throws Exception {
        Tracer.Run run = Tracer.run(List.of(TOY), "TracedToy", "gradePercent",
                (Object) new int[]{1, 1, -1, 0});
        // 2 correct out of 3 answered — the buggy denominator, faithfully reproduced. Tracing must
        // not change the answer, including when the answer is wrong.
        assertEquals(66, run.result());
        assertFalse(run.trace().isEmpty(), "the run must have been observed");
    }

    @Test
    void aVariableTimelineIsALookupNotAnInference() throws Exception {
        Tracer.Run run = Tracer.run(List.of(TOY), "TracedToy", "gradePercent",
                (Object) new int[]{1, 1, -1, 0});
        Trace t = run.trace();

        List<Trace.Obs> answered = t.values("answered");
        assertFalse(answered.isEmpty(), () -> "expected to observe `answered`; saw " + t.names());

        // Values are DATA, not reprs — this is the whole point of trace version 2, and it is what
        // lets a claim do arithmetic instead of string matching.
        assertEquals(3L, t.last("answered").value());
        assertEquals(2L, t.last("correct").value());
        assertEquals(66L, t.last("pct").value());
    }

    @Test
    void theArgumentsAVariableArrivedWithAreRecorded() throws Exception {
        Tracer.Run run = Tracer.run(List.of(TOY), "TracedToy", "gradePercent",
                (Object) new int[]{1, 1, -1, 0});
        List<Trace.Call> calls = run.trace().calls("gradePercent");
        assertEquals(1, calls.size());
        assertEquals(List.of(1L, 1L, -1L, 0L), calls.get(0).args().get("answers"));
    }

    @Test
    void aReturnIsObserved() throws Exception {
        Tracer.Run run = Tracer.run(List.of(TOY), "TracedToy", "gradePercent",
                (Object) new int[]{1, 1, -1, 0});
        List<Trace.Return> returns = run.trace().returns("gradePercent");
        assertEquals(1, returns.size());
        assertEquals(66L, returns.get(0).value());
    }

    /**
     * The reason the feature exists.
     *
     * <p>The returned percentage is self-consistent: 2 correct out of 3 answered really is 66%. No
     * claim about the RESULT can catch this. The claim that catches it is about an internal
     * variable — and it is only writable because the variable is observable.
     */
    @Test
    void theBugIsCondemnedByItsOwnTrace() throws Exception {
        int[] answers = {1, 1, -1, 0};
        Tracer.Run run = Tracer.run(List.of(TOY), "TracedToy", "gradePercent", (Object) answers);

        // The output passes every sane check you could write about it.
        int pct = (int) run.result();
        assertTrue(pct >= 0 && pct <= 100, "the result is perfectly plausible");

        // The trace is not so forgiving: the denominator never reached the number of questions.
        long answered = (long) run.trace().last("answered").value();
        assertNotEquals(answers.length, answered,
                "this is the defect: the grade was computed over answered questions, not asked ones");
        assertEquals(3L, answered);
    }

    @Test
    void anExceptionIsObservedWithTheStateUpToTheThrow() throws Exception {
        Tracer.Run[] holder = new Tracer.Run[1];
        assertThrows(ArithmeticException.class, () ->
                holder[0] = Tracer.run(List.of(TOY), "TracedToy", "divide", 7, 0));

        // The throw propagates untouched — but the tracer still saw the frame it left.
        // (The run itself threw, so the trace is read from the sink via a second, surviving run.)
        Tracer.Run ok = Tracer.run(List.of(TOY), "TracedToy", "divide", 7, 2);
        assertEquals(35, ok.result());
        assertEquals(70L, ok.trace().last("scaled").value());
    }

    @Test
    void anUntracedProcessAnswersNeverObservedRatherThanPassingVacuously() {
        Trace empty = Trace.empty();
        assertTrue(empty.isEmpty());
        assertEquals(List.of(), empty.values("answered"));
        assertNull(empty.last("answered"));
        // This is why the trace is empty-never-null: a claim about an untraced variable must FAIL,
        // not quietly succeed because there was nothing to contradict it.
        assertEquals("answered: never observed", empty.render("answered"));
    }

    @Test
    void aVersionOneTraceIsRefusedRatherThanHalfUnderstood() {
        String v1 = "{\"e\":\"H\",\"trace_version\":1}\n{\"e\":\"L\",\"fn\":\"f\",\"at\":\"a.java:1\",\"d\":{\"x\":\"42\"}}\n";
        IllegalArgumentException e = assertThrows(IllegalArgumentException.class, () -> Trace.parse(v1));
        assertTrue(e.getMessage().contains("older tracer"), e.getMessage());
    }

    @Test
    void theTraceRoundTripsThroughItsArtifactForm() throws Exception {
        Tracer.Run run = Tracer.run(List.of(TOY), "TracedToy", "gradePercent",
                (Object) new int[]{1, 1, -1, 0});
        Trace reparsed = Trace.parse(run.trace().toJsonl());
        assertEquals(run.trace().size(), reparsed.size());
        assertEquals(run.trace().last("pct").value(), reparsed.last("pct").value());
    }

    @Test
    void locationsPointAtTheOriginalFileNotTheInstrumentedOne() throws Exception {
        Tracer.Run run = Tracer.run(List.of(TOY), "TracedToy", "gradePercent",
                (Object) new int[]{1, 1, -1, 0});
        String original = java.nio.file.Files.readString(java.nio.file.Paths.get(TOY));
        int lines = original.split("\n", -1).length;

        for (Trace.Obs o : run.trace().values("answered")) {
            assertTrue(o.at().startsWith("TracedToy.java:"), o.at());
            int line = Integer.parseInt(o.at().substring("TracedToy.java:".length()));
            // Instrumenting moves every line. A location past the end of the real file would mean
            // the trace is pointing at a file that exists nowhere on the reader's disk.
            assertTrue(line <= lines, "line " + line + " is past the end of the original " + lines + "-line file");
        }
    }
}
